File: singleton/Singleton.java
public class Singleton {
    private static class Bankaccount {
        private static volatile Bankaccount instance = null;
        private int balance;

        private Bankaccount() {
            balance = 100;
        }

        public static Bankaccount getInstance() {
            if(instance == null) {
                synchronized(Bankaccount.class) {
                    if(instance == null) {
                        instance = new Bankaccount();
                    }
                }
            }
            return instance;
        }

        public int getBalance() {
            return balance;
        }

        public void withdraw() {
            balance -= 10;
        }
    }

    private class Test implements Runnable {
        private Thread t;
        private String threadName;
        
        Test( String name) {
            threadName = name;
            System.out.println("Creating " +  threadName );
        }
        
        public void run() {
            Bankaccount ac =  Bankaccount.getInstance();
            System.out.println("The bank account is: " + System.identityHashCode(ac));
            System.out.println("The balance is: " + ac.getBalance());
            ac.withdraw();
            System.out.println("The balance is: " + ac.getBalance());
        }
        
        public void start () {
            System.out.println("Starting " +  threadName );
            if (t == null) {
                t = new Thread (this, threadName);
                t.start ();
            }
        }
    }

    public void test() {
        Test R1 = new Test( "Thread-1");
        R1.start();
        
        Test R2 = new Test( "Thread-2");
        R2.start();
    }

    public static void main(String[] args) {
        Singleton s = new Singleton();
        s.test();
    }
}

File: singleton/Singleton.py
import threading


class BankAccount():
    _instance = None
    _lock = threading.Lock()

    def __init__(self):
        self.balance = 100

    def __new__(self):
        if not self._instance:
            with self._lock:
                if not self._instance:
                    self._instance = super().__new__(self)
        return self._instance
    
    def getBalance(self)->int:
        return self.balance

    def withdraw(self):
        self.balance -= 10

def thread_test_run():
    ac = BankAccount()
    print(ac)
    print(ac.getBalance())
    ac.withdraw()
    print(ac.getBalance())

if __name__ == '__main__':
    t1 = threading.Thread(target=thread_test_run)
    t2 = threading.Thread(target=thread_test_run)
    t1.start()
    t2.start()
